src/doc/doc_base.nit

```diff
@@ -0,0 +1,219 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Base entities shared by all the nitdoc code.
+module doc_base
+
+import toolcontext
+import doc_model # FIXME maintain compatibility with old templates.
+
+# The model of a Nitdoc documentation.
+#
+# `DocModel` contains the list of the `DocPage` to be generated.
+#
+# The model is populated through `DocPhase` to be constructed.
+# It is a placeholder to share data between each phase.
+class DocModel
+
+	# `DocPage` composing the documentation.
+	#
+	# This is where `DocPhase` store and access pages to produce documentation.
+	var pages = new Array[DocPage]
+
+	# Nit `Model` from which we extract the documentation.
+	var model: Model is writable
+
+	# The entry point of the `model`.
+	var mainmodule: MModule is writable
+end
+
+# A documentation page abstraction.
+#
+# The page contains a link to the `root` of the `DocComposite` that compose the
+# the page.
+class DocPage
+
+	# Title of this page.
+	var title: String is writable
+
+	# Root element of the page.
+	#
+	# `DocPhase` access the structure of the page from the `DocRoot`.
+	var root = new DocRoot
+
+	redef fun to_s do return title
+end
+
+# `DocPage` elements that can be nested in another.
+#
+# `DocComposite` is an abstraction for everything that go in a `DocPage` like
+# sections, articles, images, lists, graphs...
+#
+# It provides base services for nesting mechanisms following the
+# *Composite pattern*.
+# The composed structure is a tree from a `DocRoot` that can be manipulated
+# recursively.
+#
+# `DocComposite` should not have direct reference to a `DocPage` so it can be
+# reused for the generation of multiple page.
```

Comment:
Oui c'est une idée qui est resté dans le comment mais pas dans le code.

Alexandre

2015-02-03 3:46 GMT+01:00 Jean Privat <notifications@github.com>:

> In src/doc/doc_base.nit
> <https://github.com/privat/nit/pull/1135#discussion_r23979099>:
>
> > +
> > +	redef fun to_s do return title
> > +end
> > +
> > +# `DocPage` elements that can be nested in another.
> > +#
> > +# `DocComposite` is an abstraction for everything that go in a `DocPage` like
> > +# sections, articles, images, lists, graphs...
> > +#
> > +# It provides base services for nesting mechanisms following the
> > +# *Composite pattern*.
> > +# The composed structure is a tree from a `DocRoot` that can be manipulated
> > +# recursively.
> > +#
> > +# `DocComposite` should not have direct reference to a `DocPage` so it can be
> > +# reused for the generation of multiple page.
>
> Je suis pas sur de trop comprendre cette partie. c'est pour pouvoir
> factoriser des morceaux de page ?
> Si oui, est-ce que l'attribut DocComposite::parent ne contrevient pas à
> cette idée vu qu'il permettrai de remonter de à la racine ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/privat/nit/pull/1135/files#r23979099>.
>